phys2bids/conversion.py

```diff
@@ -0,0 +1,403 @@
+                                                   ch=None)
+
+    # Get "first" scan from multi-file acquisitions
+    df['acq_time'] = pd.to_datetime(df['acq_time'])
+    df = df.sort_values(by='acq_time')
+    df = df.drop_duplicates(subset='filename', keep='first', ignore_index=True)
+
+    # Now back to general-purpose code
+    df = determine_scan_durations(layout, df, sub=sub, ses=ses)
+    df = df.dropna(subset=['duration'])  # limit to relevant scans
+
+    # Convert scan times to relative onsets (first scan is at 0 seconds)
+    df['onset'] = (df['acq_time'] - df['acq_time'].min())
+    df['onset'] = df['onset'].dt.total_seconds()
+    return df
+
+
+def determine_scan_durations(layout, scan_df, sub, ses=None):
+    """
+    Extract scan durations by loading fMRI files/metadata and
+    multiplying TR by number of volumes. This can be used to determine the
+    endpoints for the physio files.
+
+    Parameters
+    ----------
+    layout : bids.layout.BIDSLayout
+        Dataset layout. Used to identify functional scans and load them to
+        determine scan durations.
+    scan_df : pandas.DataFrame
+        Scans DataFrame containing functional scan filenames and onset times.
+    sub : str
+        Subject ID.
+    ses : str or None, optional
+        Session ID. If None, then no session.
+
+    Returns
+    -------
+    scan_df : pandas.DataFrame
+        Updated DataFrame with new "duration" column. Calculated durations are
+        in seconds.
+    """
+    func_files = layout.get(datatype='func', suffix='bold',
+                            extension=['.nii.gz', '.nii'],
+                            subject=sub, session=ses)
+    scan_df['duration'] = None
+    for func_file in func_files:
+        filename = func_file.path
+        if filename in scan_df['original_filename'].values:
+            n_vols = nib.load(filename).shape[3]
+            tr = func_file.get_metadata()['RepetitionTime']
+            duration = n_vols * tr
+            scan_df.loc[scan_df['original_filename'] == filename, 'duration'] = duration
+        else:
+            LGR.info('Skipping {}'.format(op.basename(filename)))
+    return scan_df
+
+
+def extract_physio_onsets(trigger_timeseries, freq, threshold=0.5):
+    """
+    Collect onsets from physio file, both in terms of seconds and time series
+    indices.
+
+    Parameters
+    ----------
+    trigger_timeseries : 1D numpy.ndarray
+        Trigger time series.
+    freq : float
+        Frequency of trigger time series, in Hertz.
+    threshold : float, optional
+        Threshold to apply to binarize trigger time series.
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        DataFrame with one row for each trigger period and three columns:
+        onset (in seconds), index (onsets in index of time series array),
+        and duration (in seconds).
+    """
+    samplerate = 1. / freq
+    scan_idx = np.where(trigger_timeseries > 0)[0]
```

Comment:
This requires the trigger to be binarized already.

Reply:
Ok, I have to say I didn't have a look at the code yet, but what do you mean by this?
In which format do you need the trigger representation? Do you need a "block" timeseries that is `True`/`1` when the sequence is going on and `False`/`0` when it's not?

Reply:
Currently, it does require those blocks, but I think we can make it flexible enough to work with other formats.

Truthfully, what I really need is onsets and durations of trigger block periods in terms of seconds and array indices, so we could move the step that actually gets those to outside of the workflow if that would be easier.

Reply:
Ok, gotcha. The onset of the trigger block is a property of the classes we use throughout the pipeline (I don't remember its name now but it's there!). At the moment it's an integer because we didn't need one for each multirun block, but we can adapt it to be a dictionary and contain all of them - although that also needs to adapt the code in `split4phys`. The duration can be obtained by the `ntp*tr` variables, that are the number of timepoints expected and their length. We can add those as properties of the class as well if it helps.

Alternatively, we can add those "timeblocks" as either the last channel in the timeseries, or as another property of the class to keep the timeseries "clean". We would probably need to adapt the code in `split4phys` anyway to do create that timeblock, but it's also a good way to go IMHO.

Reply:
> Alternatively, we can add those "timeblocks" as either the last channel in the timeseries, or as another property of the class to keep the timeseries "clean".

I like the idea of a binarized version of the trigger time series, perhaps in a separate property. That way I could just assume that this property contains a boolean array and work from there. We just need to ensure that the binarization procedure works for both block-like and variable trigger signals.

Reply:
I've opened #278 about this. It will need to be address before this can be merged, and I think it merits a separate PR given that handling variable triggers is outside my area of expertise.